src/Windows/Avalonia.Win32/OleDataObjectHelper.cs

```diff
@@ -383,6 +396,76 @@ public unsafe static uint WriteDataToHGlobal(IDataTransfer dataTransfer, DataFor
         return DV_E_FORMATETC;
     }
 
+    public unsafe static uint WriteDataToGdi(IDataTransfer dataTransfer, DataFormat format, ref IntPtr hGlobalBitmap)
+    {
+        if (ClipboardFormatRegistry.HBitmapDataFormat.Equals(format))
+        {
+            var bitmap = dataTransfer.TryGetValue(DataFormat.Bitmap);
+            if (bitmap != null)
+            {
+                var pixelSize = bitmap.PixelSize;
+                var stride = ((bitmap.Format?.BitsPerPixel ?? 0) / 8) * pixelSize.Width;
+                var buffer = new byte[stride * pixelSize.Height];
+                fixed (byte* bytes = buffer)
+                {
+                    bitmap.CopyPixels(new PixelRect(pixelSize), (IntPtr)bytes, buffer.Length, stride);
+
+                    IntPtr hdc = IntPtr.Zero, compatDc = IntPtr.Zero, hbitmap = IntPtr.Zero;
+                    try
+                    {
+                        hdc = GetDC(IntPtr.Zero);
+                        if (hdc == IntPtr.Zero)
+                            return DV_E_FORMATETC;
+
+                        compatDc = CreateCompatibleDC(hdc);
+                        if (compatDc == IntPtr.Zero)
+                            return DV_E_FORMATETC;
+
+                        hbitmap = CreateCompatibleBitmap(compatDc, pixelSize.Width, pixelSize.Height);
+                        if (hbitmap == IntPtr.Zero)
+                            return DV_E_FORMATETC;
+
+                        SelectObject(compatDc, hbitmap);
+
+                        var bitmapInfoHeader = new BITMAPINFOHEADER()
+                        {
+                            biWidth = pixelSize.Width,
+                            biHeight = pixelSize.Height,
+                            biPlanes = 1,
+                            biBitCount = (ushort)(bitmap.Format?.BitsPerPixel ?? 0),
+                            biCompression = 0,
+                            biSizeImage = (uint)(pixelSize.Width * 4 * Math.Abs(pixelSize.Height))
```

Comment:
You are assuming pixel is 4 bytes here but use any BitsPerPixel value you get from the bitmap. These should be made to match. Some of the formats, and especially the ones with transparency cannot be supported using BITMAPINFOHEADER, you need V5 (or V4) header. The main benefit of using HBITMAP is that it can support more pixel formats than CF_DIB.